opm/simulators/linalg/bda/BILU0.cpp

```diff
@@ -0,0 +1,381 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/common/ErrorMacros.hpp>
+
+#include <opm/simulators/linalg/bda/BdaSolver.hpp>
+#include <opm/simulators/linalg/bda/BILU0.hpp>
+#include <opm/simulators/linalg/bda/Reorder.hpp>
+
+
+namespace bda
+{
+
+    using Opm::OpmLog;
+
+    // define 'second' as 'BdaSolver<>::second', this allows usage of the second() function for timing
+    // typedefs cannot handle templates
+    const auto second = BdaSolver<>::second;
+
+    template <unsigned int block_size>
+    BILU0<block_size>::BILU0(bool level_scheduling_, bool graph_coloring_, int verbosity_) :
+        verbosity(verbosity_), level_scheduling(level_scheduling_), graph_coloring(graph_coloring_)
+    {
+        if (level_scheduling == graph_coloring) {
+            OPM_THROW(std::logic_error, "Error, either level_scheduling or graph_coloring must be true, not both\n");
+        }
+    }
+
+    template <unsigned int block_size>
+    BILU0<block_size>::~BILU0()
+    {
+        delete[] invDiagVals;
+        delete[] diagIndex;
+        delete[] rowsPerColor;
+        freeBlockedMatrix(&LUMat);
+        freeBlockedMatrix(&LMat);
+        freeBlockedMatrix(&UMat);
+        delete[] toOrder;
+        delete[] fromOrder;
+        freeBlockedMatrix(&rMat);
+    }
+
+    template <unsigned int block_size>
+    bool BILU0<block_size>::init(BlockedMatrix *mat)
+    {
+        const unsigned int bs = block_size;
+        double t1 = 0.0, t2 = 0.0;
+        BlockedMatrix *CSCmat = nullptr;
+
+        this->N = mat->Nb * block_size;
+        this->Nb = mat->Nb;
+        this->nnz = mat->nnzbs * block_size * block_size;
+        this->nnzbs = mat->nnzbs;
+
+        toOrder = new int[N];
+        fromOrder = new int[N];
+        if (level_scheduling) {
+            CSCmat = new BlockedMatrix();
+            CSCmat->Nb = Nb;
+            CSCmat->nnzbs = nnzbs;
+            CSCmat->nnzValues = new double[nnzbs * bs * bs];
+            CSCmat->colIndices = new int[nnzbs];
+            CSCmat->rowPointers = new int[Nb + 1];
+            if(verbosity >= 3){
+                t1 = second();
+            }
+            bcsr_to_bcsc<block_size>(mat->nnzValues, mat->colIndices, mat->rowPointers, CSCmat->nnzValues, CSCmat->colIndices, CSCmat->rowPointers, mat->Nb);
+            if(verbosity >= 3){
+                t2 = second();
+                std::ostringstream out;
+                out << "BILU0 convert CSR to CSC: " << t2 - t1 << " s";
+                OpmLog::info(out.str());
+            }
+        }
+
+        if(verbosity >= 3){
+            t1 = second();
+        }
+        rMat = allocateBlockedMatrix<block_size>(mat->Nb, mat->nnzbs);
+        LUMat = soft_copyBlockedMatrix(rMat);
```

Comment:
I agree that the memory ownership is messy here. It is needed, however, because the rMat and LUMat share the same sparsity pattern, while having different nnzValues arrays (in line 120 a new double array is assigned to LUMat->nnzValues). This method seemed preferable to copying the colIndices and rowPointers arrays from rMat into LUMat, when none of those arrays of either of the matrices are modified after this point.

Reply:
Well, since you seem to like C-arrays, how about this:
```
#include<memory>
#include<iostream>
#include<cassert>
template<int BS>
struct BlockedMatrix{
     BlockedMatrix(int N, int nnz)
         : nnzValues(new double[nnz*BS*BS]), colIndices(new int[nnz*BS*BS]),
           rowPointers(new int[N+1]), Nb(N), nnzbs(nnz)
     {}
     /// \brief soft copy
     BlockedMatrix(const BlockedMatrix& M)
     : nnzValues(new double[M.Nb*BS*BS]),
       colIndices(M.colIndices), rowPointers(M.rowPointers),
      Nb(M.Nb), nnzbs(M.nnzbs)
     {}
    // proper deallocation is done by the smart pointers.
    std::unique_ptr<double[]> nnzValues;
    std::shared_ptr<int[]> colIndices;
    std::shared_ptr<int[]> rowPointers;
    int Nb;
    int nnzbs;
};

int main()
{
    BlockedMatrix<3> bm(10, 1000);
    BlockedMatrix<3> bm1(bm);
    assert(bm.colIndices.get()==bm1.colIndices.get());
    assert(bm.rowPointers.get()==bm1.rowPointers.get());
    assert(bm.nnzValues.get()!=bm1.nnzValues.get());
    std::cout<<bm1.nnzValues[3]<<" "<<bm1.colIndices.get()[3]<<std::endl;
    
}
```
This has all your memory management and soft-copy code in it. It does not copy what it should not. Is rather short.

Reply:
But maybe the struct has to be pure C?

Reply:
The struct and the whole reordering was originally C. That's why there are still quite a few raw pointers. We don't mind using `std::vector`, but this seems safe and consise.

Reply:
I'm have a problem implementing this solution. `openclSolver` has a `BlockedMatrix mat`, which represents the original matrix. The `mat->nnzValues` points to the first double of the `BCRSMatrix` from the `ISTLSolver`, because `mat->nnzValues.reset(vals)` in `openclSolver::initialize()`. By calling `matrix.reset(vals)` in `openclSolver::update_system`, the previous array is deleted, which belongs to the `BCRSMatrix`.
Changing the `nnzValues`, `colIndices` and `rowPointers` to `std::vector` might solve some problems, but a `double *` cannot be assigned to a `std::vector` without copying all data.
We can also go back to using raw pointers, and letting the destructor of `BlockedMatrix` delete all data.

Reply:
It does not only seem to be the nnzvalues that are reused but also rowstart and column indices.

The problem is there because cusparseSolverbackend is CUDA code compiled with NVCC and cannot use DUNE classes but needs plain-C. As cusparseSolverBackend shares the same interface as openclSolverBackend this limitation is passed on to the openCl code. Is that right?

An alternative might be to introduce another layer for cuda that gets BCRSMatrix and passes the pointers onto the solve_system cuda kernel. This new layer (with BCRSMatrix) becomes the interface. I guess OpenCL could work with only the interface of solve_system taking the BCRSMatrix. BlockedMatrix would get a constructor taking a BCRSMatrix.

Or one simply stores two backends in BdaBridge, one for openCl and one for CUDA and chooses the correct with an if at the two places. Thus the openCl one could use a BCRSMatrix.

But maybe I am now bothering you too much.

Reply:
We assume the sparsity pattern stays contant over the course of the simulation. That's why `update_system()` only updates the nonzeroes. The pointer to the nonzeroes cannot be reused since the BCRSMatrix is reset every linear solve.
`cusparseSolver` and `openclSolver` are both `BdaSolver`s, which take raw pointers. These are 'generated' in `BdaBridge`. The could both use `BCRSMatrix`, it's the `cudaMemcpy()`, `cusparseDot()`, `clWriteBuffer()`, etc that require raw pointers.
NVCC does accept a `--std=c++[11|14|17]` flag, but I have not tested that yet.
If `openclSolver` uses a BCRSMatrix, then the underlying should also do that. I don't think a BCRSMatrix can be reordered easily without rebuilding the whole thing.

Reply:
I do have a working version (up to Report step 68) with raw pointers:
```
template<int BS>
struct BlockedMatrix{
    /// Allocate BlockedMatrix and data arrays with given sizes
    BlockedMatrix(int Nb_, int nnzbs_)
    : nnzValues(new double[nnzbs_*BS*BS]),
      colIndices(new int[nnzbs_*BS*BS]),
      rowPointers(new int[Nb_+1]),
      Nb(Nb_),
      nnzbs(nnzbs_),
      deleteNnzs(true),
      deleteSparsity(true)
    {}
    /// \brief soft copy the sparsity pattern, allocate for nonzeroes if indicated
    BlockedMatrix(const BlockedMatrix& M)
    : nnzValues(new double[M.nnzbs*BS*BS]),
      colIndices(M.colIndices),
      rowPointers(M.rowPointers),
      Nb(M.Nb),
      nnzbs(M.nnzbs),
      deleteNnzs(true),
      deleteSparsity(false)
    {}
    /// \brief create BlockedMatrix pointing to existing arrays
    BlockedMatrix(int Nb_, int nnzbs_, double *nnzValues_, int *colIndices_, int *rowPointers_)
    : nnzValues(nnzValues_),
      colIndices(colIndices_),
      rowPointers(rowPointers_),
      Nb(Nb_),
      nnzbs(nnzbs_),
      deleteNnzs(false),
      deleteSparsity(false)
    {}

    ~BlockedMatrix(){
        if (deleteNnzs) {
            delete[] nnzValues;
        }
        if (deleteSparsity) {
            delete[] colIndices;
            delete[] rowPointers;
        }
    }

    double *nnzValues;
    int *colIndices;
    int *rowPointers;
    int Nb;
    int nnzbs;
    bool deleteNnzs;
    bool deleteSparsity;
};
```

Reply:
Great. I mark this as resolved then.